contracts/cw721-base/src/contract.rs

```diff
@@ -14,442 +12,609 @@ use cw721::{
+        set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+        let info = ContractInfoResponse {
+            name: msg.name,
+            symbol: msg.symbol,
+        };
+        self.contract_info.save(deps.storage, &info)?;
+        let minter = deps.api.addr_validate(&msg.minter)?;
+        self.minter.save(deps.storage, &minter)?;
+        Ok(Response::default())
+    }
 
-pub fn execute_send_nft(
-    deps: DepsMut,
-    env: Env,
-    info: MessageInfo,
-    contract: String,
-    token_id: String,
-    msg: Binary,
-) -> Result<Response, ContractError> {
-    // Transfer token
-    _transfer_nft(deps, &env, &info, &contract, &token_id)?;
-
-    let send = Cw721ReceiveMsg {
-        sender: info.sender.to_string(),
-        token_id: token_id.clone(),
-        msg,
-    };
-
-    // Send message
-    Ok(Response::new()
-        .add_message(send.into_cosmos_msg(contract.clone())?)
-        .add_attribute("action", "send_nft")
-        .add_attribute("sender", info.sender)
-        .add_attribute("recipient", contract)
-        .add_attribute("token_id", token_id))
-}
+    pub fn execute(
+        &self,
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        msg: ExecuteMsg<T>,
+    ) -> Result<Response<C>, ContractError> {
+        match msg {
+            ExecuteMsg::Mint(msg) => self.execute_mint(deps, env, info, msg),
+            ExecuteMsg::Approve {
+                spender,
+                token_id,
+                expires,
+            } => self.execute_approve(deps, env, info, spender, token_id, expires),
+            ExecuteMsg::Revoke { spender, token_id } => {
+                self.execute_revoke(deps, env, info, spender, token_id)
+            }
+            ExecuteMsg::ApproveAll { operator, expires } => {
+                self.execute_approve_all(deps, env, info, operator, expires)
+            }
+            ExecuteMsg::RevokeAll { operator } => {
+                self.execute_revoke_all(deps, env, info, operator)
+            }
+            ExecuteMsg::TransferNft {
+                recipient,
+                token_id,
+            } => self.execute_transfer_nft(deps, env, info, recipient, token_id),
+            ExecuteMsg::SendNft {
+                contract,
+                token_id,
+                msg,
+            } => self.execute_send_nft(deps, env, info, contract, token_id, msg),
+        }
+    }
 
-pub fn _transfer_nft(
-    deps: DepsMut,
-    env: &Env,
-    info: &MessageInfo,
-    recipient: &str,
-    token_id: &str,
-) -> Result<TokenInfo, ContractError> {
-    let mut token = tokens().load(deps.storage, &token_id)?;
-    // ensure we have permissions
-    check_can_send(deps.as_ref(), env, info, &token)?;
-    // set owner and remove existing approvals
-    token.owner = deps.api.addr_validate(recipient)?;
-    token.approvals = vec![];
-    tokens().save(deps.storage, &token_id, &token)?;
-    Ok(token)
-}
+    pub fn execute_mint(
```

Comment:
Now there is the clue of #391 - the idea is to move all of those functions to the trait like:

```rust
trait Cw721<T> where T: Serialize + DeserializeOwned + Clone {
  type Err: ToString;
  type C: Serialize + DeserializeOwned + Clone;

  fn mint(&self, deps: DepsMut, _env: Env, MessageInfo, msg: MingMsg<T>) -> Result<Response<Self::C>, Self::Err>;
}
```

Then the trait is implemented by contract itself, but may also be implemented by any other contract which also react to this contract msg.

Also I find this naming very verbose - basically `execute` sounds like `do action` so it feels like `mint` is just expressive enough. I would personally avoid prefixing functions if the name is not ambiguous, and leave them matching the message name - but this is side note.

Reply:
Yes, I was doing the minimal subset here. I will pull out traits, sounds good.

Reply:
> type C: Serialize + DeserializeOwned + Clone

Note this contradicts what I just did in https://github.com/CosmWasm/cw-plus/pull/445/commits/b32b21e6919cb15d49875ec5b6df0b5930562a8a which makes C generic on the method (you fix it for the contract when implementing the trait).

If we remove Self::C and keep Self::Error I can see this coming along

Reply:
I started a bit of this in 19803f21 to get first feedback.

Note that Mint is not in the generic cw721 standard, so we have the case already of the contract extending the default. This is more when you try to move execute into the message type.

Reply:
> I actually don't see any sense in making C generic in methods - its semantic is to be always the same for given T. it makes no sense for having it generic to be decided on call moment, and more important - there is no real way to elide its type there!

T is the extra data added to TokenInfo when storing it.
C is custom message type from Response<C>.
These have nothing to do with each other.

But I do agree to putting this back into a trait/struct bound not a method bound

Reply:
Yup, I figured it out bit later. It should definitely be on trait bound, probably not necessary for a struct itself (you can implement generic struct with generics which your struct is not generic over), but it is a detail (I elaborated in different comment).